lib/Scheduling/SimplexSchedulers.cpp

```diff
@@ -842,6 +878,225 @@ LogicalResult SharedOperatorsSimplexScheduler::schedule() {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// ModuloSimplexScheduler
+//===----------------------------------------------------------------------===//
+
+LogicalResult ModuloSimplexScheduler::MRT::enter(Operation *op,
+                                                 unsigned timeStep) {
+  auto opr = *sched.prob.getLinkedOperatorType(op);
+  auto lim = *sched.prob.getLimit(opr);
+  assert(lim > 0);
+
+  auto &revTab = reverseTables[opr];
+  assert(!revTab.count(op));
+
+  unsigned slot = timeStep % sched.parameterT;
+  auto &cell = tables[opr][slot];
+  if (cell.size() < lim) {
+    cell.insert(op);
+    revTab[op] = slot;
+    return success();
+  }
+  return failure();
+}
+
+void ModuloSimplexScheduler::MRT::release(Operation *op) {
+  auto opr = *sched.prob.getLinkedOperatorType(op);
+  auto &revTab = reverseTables[opr];
+  auto it = revTab.find(op);
+  assert(it != revTab.end());
+  tables[opr][it->second].erase(op);
+  revTab.erase(it);
+}
+
+bool ModuloSimplexScheduler::fillObjectiveRow(SmallVector<int> &row,
+                                              unsigned obj) {
+  switch (obj) {
+  case OBJ_LATENCY:
+    // Minimize start time of user-specified last operation.
+    row[startTimeLocations[startTimeVariables[lastOp]]] = 1;
+    return true;
+  case OBJ_AXAP:
+    // Minimize sum of start times of all-but-the-last operation.
+    for (auto *op : getProblem().getOperations())
+      if (op != lastOp)
+        row[startTimeLocations[startTimeVariables[op]]] = 1;
+    return false;
+  default:
+    llvm_unreachable("Unsupported objective requested");
+  }
+}
+
+void ModuloSimplexScheduler::updateMargins() {
+  // Assumption: current secondary objective is "ASAP".
+  // Negate the objective row once to effectively maximize the sum of start
+  // times, which yields the "ALAP" times after solving the tableau. Then,
+  // negate it again to restore the "ASAP" objective, and store these times as
+  // well.
+  for (auto *axapTimes : {&alapTimes, &asapTimes}) {
+    multiplyRow(OBJ_AXAP, -1);
+    // This should not fail for a feasible tableau.
+    assert(succeeded(restoreDualFeasibility()) && succeeded(solveTableau()));
```

Comment:
I have been bit by this before: the `assert` macro shouldn't be used for side-effecting computations, because this line simply doesn't execute when the compiler is built with assertions disabled. In order to avoid this pitfall, and also avoid unused variable warnings when assertions are disabled, I'd recommend a pattern like:

```suggestion
    auto feasible = restoreDualFeasibility();
    auto solvable = solveTableau();
    assert(succeeded(feasible) && succeeded(solvable));
    (void)feasible;
    (void)solvable;
```

Reply:
🤦‍♂️ Thanks for spotting this!